### PR TITLE
Sl input tests

### PR DESCRIPTION
**Pull request by azizpunjani**

### CODE REVIEW

File: addon/components/sl-input.js [hunk 1]
```diff
@@ -26,6 +26,8 @@ export default Ember.Component.extend( InputBased, TooltipEnabled, {
     /** @type {Object} */
     layout,
 
+    /** @type {String} */
+    dataTrigger: 'focus',
     // -------------------------------------------------------------------------
```
Comment (notmessenger) [inline addon/components/sl-input.js]: Add extra line(s) before this line

File: addon/mixins/sl-tooltip-enabled.js [hunk 2]
```diff
@@ -85,7 +93,7 @@ export default Ember.Mixin.create({
 
         // First-time rendering
         if ( 'undefined' === Ember.typeOf( originalTitle ) ) {
-            this.set( 'data-toggle', 'popover' );
+            this.set( 'dataToggle', 'popover' );
```
Comment (notmessenger) [inline addon/mixins/sl-tooltip-enabled.js]: If `popover` is a string then is the `dataToggle` property's type incorrect?

File: addon/mixins/sl-tooltip-enabled.js [hunk 3]
```diff
@@ -28,11 +29,18 @@ export default Ember.Mixin.create({
...
-     * 'data-toggle' attribute for use in template binding
+     * dataToggle property
      *
      * @type {?Boolean}
      */
-    'data-toggle': null,
+    dataToggle: null,
+
+    /**
+     * dataTrigger
```
Comment (notmessenger) [inline addon/mixins/sl-tooltip-enabled.js]: Improve the description

File: tests/unit/mixins/sl-tooltip-enabled-test.js [hunk 4]
```diff
@@ -156,9 +156,9 @@ test( 'enablePopover() - Renders popover', function( assert ) {
     subject.enablePopover();
 
     assert.equal(
-        subject.get( 'data-toggle' ),
+        subject.get( 'dataToggle' ),
         'popover',
-        '"data-toggle" has correct value'
+        '"dataToggle" has correct value'
     );
```
Comment (notmessenger) [inline tests/unit/mixins/sl-tooltip-enabled-test.js]: Add a test for `dataToggle to ensure that default value is `null` and doesn't get changed to something else, like `focus`.

File: tests/unit/components/sl-input-test.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
+import ember from 'ember';
+import { moduleForComponent, test } from 'ember-qunit';
+
+moduleForComponent( 'sl-input', 'Unit | Component | sl input', {
+    unit: true
+});
+
+
```
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: Don't need a test for the `enter` action?

File: tests/unit/components/sl-input-test.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+                );
+
+                done();
+            }
+        }
+    });
+
+    this.$( 'input' ).trigger( 'blur' );
+});
+
```
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: Add a test for default values, specifically `clickToEdit`, `isTypeaheadSetup`, `suggestionNamePath` and `type`

File: tests/unit/components/sl-input-test.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        helpText: helpText
+    });
+
+    assert.equal(
+        this.$( '.help-block' ).text().trim(),
+        helpText
+    );
+});
+
+test( 'Label text is displayed', function( assert ) {
```
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: There should be a negation test to this one.  if the `{{#if label}}` condition in the template is removed this test will still pass, when what we are really wanting to assert is that the label is only rendered when the `label` property is populated.

File: tests/unit/components/sl-input-test.js [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        label: labelText
+    });
+
+    assert.equal(
+        this.$( '.control-label' ).text().trim(),
+        labelText
+    );
+});
+
+test( 'Optional property displays optional label', function( assert ) {
```
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: Probably throw some tests for the negation of this one and `required` property in the `label` one(s) for the same reason left in the comment on the `label` test.

File: tests/unit/components/sl-input-test.js [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        placeholder: 'placeholder'
+    });
+
+    assert.equal(
+        this.$( 'input' ).prop( 'placeholder' ),
+        placeholderText
+    );
+});
+
+test( 'Popover is displayed and has the correct title and content', function( assert ) {
```
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: Is the un-displaying of the popover when the element is unfocused tested in any of the other tests (such as the mixin)?  If not, such a tests should exist here.

File: tests/unit/components/sl-input-test.js [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    this.subject({
+        readonly: true
+    });
+
+    assert.ok(
+        this.$( 'input' ).prop( 'readonly' )
+    );
+});
+
+test( 'Disabled property, disables the input', function( assert ) {
```
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: Have already tested this in this file.

File: tests/unit/components/sl-input-test.js [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    const colors =  [
+        'Black',
+        'Yellow'
+    ];
+
+    const component = this.subject({
+        suggestions: colors
+    });
+
+    this.render();
```
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: This line is not needed as the `this.$()` call in line 190 accomplishes the same thing.

**Comment by notmessenger** [comment]
Should test that registered events are unregistered.

**Comment by notmessenger** [comment]
There are no tests for `getInput()`.

**Comment by notmessenger** [comment]
`inputClass()` is indirectly tested by confirming that when `clickToEdit` is true the `click-to-edit` class is applied.  I did not see any tests for the `suggestions` condition, as well as this method should be tested independently and directly.

**Comment by notmessenger** [comment]
https://github.com/softlayer/sl-ember-components/blob/develop/addon/templates/components/sl-input.hbs#L2 and https://github.com/softlayer/sl-ember-components/blob/develop/addon/templates/components/sl-input.hbs#L16 should be replaced with the pattern found in the `sl-date-picker` component.  There should also be tests for this value.

File: addon/mixins/sl-tooltip-enabled.js (hunk 3 above)
Comment (azizpunjani) [inline addon/mixins/sl-tooltip-enabled.js]: :white_check_mark:

File: addon/mixins/sl-tooltip-enabled.js (hunk 2 above)
Comment (azizpunjani) [inline addon/mixins/sl-tooltip-enabled.js]: :white_check_mark:

File: addon/components/sl-input.js (hunk 1 above)
Comment (azizpunjani) [inline addon/components/sl-input.js]: :white_check_mark:

File: tests/unit/components/sl-input-test.js (hunk 5 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: As discussed, removing the enter action as it's not used or referenced in the docs.

File: tests/unit/components/sl-input-test.js (hunk 6 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: :white_check_mark:

File: tests/unit/components/sl-input-test.js (hunk 7 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: ::white_check_mark:

File: tests/unit/components/sl-input-test.js (hunk 8 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: :white_check_mark:

File: tests/unit/components/sl-input-test.js (hunk 10 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: :white_check_mark:

File: tests/unit/components/sl-input-test.js (hunk 11 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: :white_check_mark:

File: tests/unit/components/sl-input-test.js (hunk 10 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: :white_check_mark:

File: tests/unit/components/sl-input-test.js (hunk 11 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: :white_check_mark:

**Comment by notmessenger** [comment]
@azizpunjani Build is failing

**Comment by azizpunjani** [comment]
Still working on this.

File: tests/unit/mixins/sl-tooltip-enabled-test.js (hunk 4 above)
Comment (azizpunjani) [inline tests/unit/mixins/sl-tooltip-enabled-test.js]: :white_check_mark:

**Comment by azizpunjani** [comment]
@notmessenger PR is ready for review.

File: tests/unit/components/sl-input-test.js (hunk 6 above)
Comment (notmessenger) [inline tests/unit/components/sl-input-test.js]: Default value for `type` is not asserted in the "Default values are correct" test

File: tests/unit/mixins/sl-tooltip-enabled-test.js [hunk 12]
```diff
@@ -36,8 +36,31 @@ test( 'Successfully mixed', function( assert ) {
...
+    );
+
+    assert.strictEqual(
+        subject.get( 'title' ),
+        null,
+        'title is null'
+    );
+});
+
+window.QUnit.skip( 'enable() - observes correct properties', function() {
```
Comment (notmessenger) [inline tests/unit/mixins/sl-tooltip-enabled-test.js]: replace `window.QUnit.skip()` with skip()

File: tests/unit/components/sl-input-test.js (hunk 6 above)
Comment (azizpunjani) [inline tests/unit/components/sl-input-test.js]: :white_check_mark:

**Comment by azizpunjani** [comment]
@notmessenger this PR is ready for review, I have made requested modifications.
